rust/kernel/of.rs

```diff
@@ -4,72 +4,84 @@
 //!
 //! C header: [`include/linux/of_*.h`](../../../../include/linux/of_*.h)
 
-use alloc::boxed::Box;
+use crate::{bindings, c_types, str::CStr};
 
-use crate::{
-    bindings, c_types,
-    error::{Error, Result},
-    str::CStr,
-    types::PointerWrapper,
-};
-
-use core::mem::transmute;
-
-type InnerTable = Box<[bindings::of_device_id; 2]>;
+use core::ptr;
 
 /// Wraps a kernel Open Firmware / devicetree match table.
 ///
-/// Rust drivers may create this structure to match against devices
+/// Rust drivers may use this trait to match against devices
 /// described in the devicetree.
-///
-/// The ['PointerWrapper'] trait provides conversion to/from a raw pointer,
-/// suitable to be assigned to a `bindings::device_driver::of_match_table`.
-///
-/// # Invariants
-///
-/// The final array element is always filled with zeros (the default).
-pub struct OfMatchTable(InnerTable);
+pub trait OfMatchTable {
```

Comment:
I don't think trait is correct abstraction here. It should look more like `CStr` and `CBoundedStr`.

Reply:
You think so? All an `OfMatchTable` consumer (i.e. a driver) needs, is a way to retrieve a static-lifetime pointer, right?

Reply:
> I suspect the strange thing you observe is related to the usage of incomplete features.

I haven't seen any strange things yet - I was just wondering if the array behaviour could be different for platforms I cannot test. Good to hear that it's correct, and no `repr` is required.

Reply:
@nbdd0121 can you elaborate why trait isn't the correct abstraction here?

Reply:
Conceptually `OfMatchTable` is to `ConstOfMatchTable` as slice is to array. Obvious slice is not abstracted as a trait that all arrays implement.

I think what you should do here is to have a
```rust
/// INVARIANT: end with a sentinel
pub struct OfMatchTable([bindings::of_device_id]);
```
or maybe
```rust
/// INVARIANT: point to a slice that end with a sentinel
pub struct OfMatchTable(bindings::of_device_id);
```

and then have `ConstOfMatchTable` implement `Deref<Target=OfMatchTable>`. This way you'll only need to accept a `&'static OfMatchTable`.

Reply:
I agree with you that, in general, a `String/&str` type abstraction pair is preferable over a `dyn trait`. However, I can't find a way to implement your suggestion without using unsafe code:
```rust
/// Wraps a kernel Open Firmware / devicetree match table.
#[repr(transparent)]
pub struct OfMatchTable([bindings::of_device_id]);

impl<const N: usize> Deref for ConstOfMatchTable<N> {
    type Target = OfMatchTable;

    fn deref(&self) -> &OfMatchTable {
        unsafe { &*(&self.table[..] as *const [bindings::of_device_id] as *const OfMatchTable) }
    }
}
```
That's really unfortunate, because the `dyn trait` solution uses only safe Rust. Even coercing the `*const` reference is fully safe, as the pointer has `'static` lifetime.

Would I be able to convince you to prefer the slightly more unseemly, yet safe, abstraction, over the pretty-yet-unsafe version?

Reply:
I think the `kernel` crate should aim to provide ergonomic abstractions, accepting *internal* usage of unsafe. What `kernel` crate is to the drivers is pretty similar to what `std` is to userspace Rust programs, and it is ubiquitous in `std` to perform this sort of cast (e.g. https://doc.rust-lang.org/stable/src/std/path.rs.html#1791-1793).

I think we should stop demonizing `unsafe`s and avoiding them at all cost. If usage of `unsafe` is helpful, then use it and just justify why it's safe with safety comments. It's not "pretty-yet-unsafe" abstraction; it is a pretty-and-safe abstraction, with a justified use of `unsafe` operation internally.

Reply:
> If usage of `unsafe` is helpful, then use it and just justify why it's safe with safety comments. It's not "pretty-yet-unsafe" abstraction; it is a pretty-and-safe abstraction, with a justified use of `unsafe` operation internally.

This implies that we are superhuman, and never make mistakes :)

Let's start by determining whether the following is safe. The table is stored as an array[N] + sentinel. I am hoping that the `repr(C)` will pack the two, so that the whole behaves as an array[N + 1]. Is that a correct assumption?
```rust
#[repr(C)]
pub struct ConstOfMatchTable<const N: usize> {
    table: [bindings::of_device_id; N],
    sentinel: bindings::of_device_id,
}
```
Then, the `OfMatchTable` type (that can only exist as a reference) stores the table as a slice. It has no way to store the sentinel:
```rust
#[repr(transparent)]
pub struct OfMatchTable([bindings::of_device_id]);
```
If we now take a pointer to the first element of the slice, is that equivalent to a pointer to array[N] + sentinel?
```rust
impl OfMatchTable {
    /// Return the table as a static lifetime, sentinel-terminated C array.
    ///
    /// This is suitable to be assigned to the kernel's `of_match_table` field.
    pub fn as_ptr(&'static self) -> *const bindings::of_device_id {
        &self.0[0]
    }
}
```

Reply:
> I think we should stop demonizing `unsafe`s and avoiding them at all cost. If usage of `unsafe` is helpful, then use it and just justify why it's safe with safety comments. It's not "pretty-yet-unsafe" abstraction; it is a pretty-and-safe abstraction, with a justified use of `unsafe` operation internally.

The main reason I want rust in the kernel is its safe subset. I generally agree that when the use of unsafe brings a great usability improvement, then we should consider it. However, I think we should still avoid it -- we are all susceptible to making mistakes, having the compiler catch them for us is highly desirable, especially in the kernel crate.

Reply:
Like @wedsonaf I believe that we should avoid `unsafe` unless there's a great usability improvement. That said, I'll go with the consensus here. Wedson, you are currently reviewing the `unsafe` version. Would you like to see the safe version as well? It's [here](https://github.com/Rust-for-Linux/linux/commit/a3ce5a67ae80b1d32b250229906b2d7441a1ce9a).